src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,125 @@ public SubstitutionMatrix(final byte[] matrix) {
         baseByCode['N'][(encodedMatrixBytes[4] >> 2) & 3] = 'G';
         baseByCode['N'][(encodedMatrixBytes[4]) & 3] = 'T';
 
-        for (final byte refBase : SUBSTITUTION_BASES_UPPER) {
+        for (final SubstitutionBase refBase : BASES) {
             for (byte code = 0; code < CODES_PER_BASE; code++) {
-                codeByBase[refBase][baseByCode[refBase][code]] = code;
+                codeByBase[refBase.getBase()][baseByCode[refBase.getBase()][code]] = code;
             }
         }
     }
 
+    /**
+     * Given a reference base and a read base, find the corresponding substitution code
+     * @param refBase reference base being substituted
+     * @param readBase read base to substitute for the reference base
+     * @return code to be used for this refBase/readBase pair
+     */
+    public byte code(final byte refBase, final byte readBase) {
+        if (refBase <= 0 || Character.isLowerCase((char) refBase)) {
+            throw new IllegalArgumentException(
+                    String.format("CRAM: Attempt to generate a substitution code for invalid or lower case reference base '%c'", (char) refBase));
+        }
+        if (readBase <= 0) {
+            throw new IllegalArgumentException(
+                    String.format("CRAM: Attempt to generate a substitution code for an invalid read base value '%c'", (char) readBase));
+        }
+        return codeByBase[refBase][readBase];
+    }
+
+    /**
+     * Given a reference base and a substitution code, return the corresponding substitution base.
+     * @param refBase reference base being substituted
+     * @param code substitution code
+     * @return base to be substituted for this (refBase, code) pair
+     */
+    public byte base(final byte refBase, final byte code) {
+        if (refBase <= 0) {
+            throw new IllegalArgumentException(
+                    String.format("CRAM: Attempt to generate a substitution code for invalid reference base '%c'", (char) refBase));
+        }
+        final byte base = baseByCode[refBase][code];
+        if (base == NO_BASE) {
+            // attempt to retrieve a code for a reference base that isn't in the substitution matrix
+            throw new IllegalArgumentException(String.format("CRAM: Attempt to retrieve a substitution base for invalid base '%c'", (char) refBase));
+        }
+        return base;
+    }
+
     /**
      * Return this substitution matrix as a byte array in a form suitable for serialization.
-     * @return
+     * @return the encoded matrix in serialized form
      */
     public byte[] getEncodedMatrix() {
         return encodedMatrixBytes;
     }
 
+    @Override
+    public String toString() {
+        final int DISPLAY_SIZE = BASES_SIZE * (CODES_PER_BASE + 3) * 2;
```

Comment:
if you want you can move this into a static constant field since is in upper case... if its a local variable I would make it lowercase or simply inline the code in the ```new StringBuilder(...)``` statement.